feat: [IA-241] Implements new Payment Transaction error screen

CrisTofani opened the pull request:
## Short description
This PR reworks the Transaction Error screen shown after the activation of a payment in case of errors. This PR handles the errors implemented by the backend in pagopa/io-backend#828

## How to test
Change the error response on `io-dev-api-server` [here](https://github.com/pagopa/io-dev-api-server/blob/master/src/routers/payment.ts#L44) or [here](https://github.com/pagopa/io-dev-api-server/blob/master/src/routers/payment.ts#L63) to return the newly defined error payload.

[Here](https://docs.google.com/spreadsheets/d/16VBHabKr4b3HBII-P0VuLByhWLRSU4Y3-TwYMu077MU/edit#gid=0) the map of the complete cases of errors given by the new implementation

| Error Type | Screen |
| ------ | ------- |
| **TECHNICAL** | <img src="https://user-images.githubusercontent.com/3959405/133973811-c9a23b0c-13e2-483b-8ee2-3158ee417f03.png" height="550"/> |
| **DATA** | <img src="https://user-images.githubusercontent.com/3959405/133973881-63b5fd6f-7a26-476c-ab62-019dfefbe6e2.png" height="550"/> |
| **EC** | <img src="https://user-images.githubusercontent.com/3959405/133974076-e57fadd9-100d-494b-9e48-43b575728ad7.png" height="550"/> | 
| **ONGOING** | <img src="https://user-images.githubusercontent.com/3959405/133974149-568d9b12-1de0-4f25-a0fb-3b7f1b7e6f0a.png" height="550"/> | 
| **EXPIRED** | <img src="https://user-images.githubusercontent.com/3959405/133974276-2654f4cd-bffd-4569-9846-31b19ce7027d.png" height="550"/> |
| **REVOKED** | <img src="https://user-images.githubusercontent.com/3959405/133974366-83cbb544-9f1d-4a7e-9814-4ba1a3f1522b.png" height="550"/> |
| **DUPLICATED** | <img src="https://user-images.githubusercontent.com/3959405/133989084-7fbda567-3c41-43e6-bbab-cc02eedf6aa7.png" height="550"/> |
| **UNCOVERED** | <img src="https://user-images.githubusercontent.com/3959405/133974401-69f59827-4159-42a5-948b-58d605d7a405.png" height="550"/> |

thisisjp commented:
@CrisTofani before reviewing the copy, please note there shouldn't be any header (nor back button neither title) in this page

thisisjp commented:
@CrisTofani sorry for the double comment: in the umbrella screen the description seems aligned on the left, while it should be centered—is it just a copy-paste issue?

CrisTofani commented:
> @CrisTofani before reviewing the copy, please note there shouldn't be any header (nor back button neither title) in this page

Fixed within 7cc35fb

> @CrisTofani sorry for the double comment: in the umbrella screen the description seems aligned on the left, while it should be centered—is it just a copy-paste issue?

It was a style issue on the code, thank's for catching it! Fixed within 79a3c57
<img src="https://user-images.githubusercontent.com/3959405/134001754-30c2fea4-7813-4265-861b-fb2a5d40596a.png" height="550"/>

Undermaken commented:
@CrisTofani I added the label ✋🏻  `don't merge` until [this PR](https://github.com/pagopa/io-backend/pull/828) is not merged and deployed
btw we can review this PR

CrisTofani commented:
> @CrisTofani I added the label ✋🏻 `don't merge` until [this PR](https://github.com/pagopa/io-backend/pull/828) is not merged and deployed
> btw we can review this PR

Yes, thanks! I'm waiting for news by the backend team, once this will be merged I will remove the tag

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
     text: !include wallet/why_a_fee.md
   errors:
-    PAYMENT_DUPLICATED: This notice has been already paid!
+    TECHNICAL: Apologies, there is a technical problem with this notice
```
```suggestion
    TECHNICAL: Sorry, a technical problem has occurred with this notice
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
   errors:
-    PAYMENT_DUPLICATED: This notice has been already paid!
+    TECHNICAL: Apologies, there is a technical problem with this notice
+    DATA: Apologies, the codes entered are incorrect
```
```suggestion
    DATA: Sorry, the payment information is incorrect
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
+    DATA: Apologies, the codes entered are incorrect
+    EC: Apologies, the Creditor Body is having trouble responding
+    PAYMENT_ONGOING: There is already a payment in progress, please try again in a few minutes.
+    ONGOING_SUBTITLE: If too much time has passed, please let us know!
```
```suggestion
    ONGOING_SUBTITLE: If it has been too long, please let us know!
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
+    EC: Apologies, the Creditor Body is having trouble responding
+    PAYMENT_ONGOING: There is already a payment in progress, please try again in a few minutes.
+    ONGOING_SUBTITLE: If too much time has passed, please let us know!
+    REVOKED: Apologies, the Creditor Body has withdrawn this notice
```
```suggestion
    REVOKED: Sorry, the Payee has revoked this notice
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
+    PAYMENT_ONGOING: There is already a payment in progress, please try again in a few minutes.
+    ONGOING_SUBTITLE: If too much time has passed, please let us know!
+    REVOKED: Apologies, the Creditor Body has withdrawn this notice
+    EXPIRED: Apologies, the notice has expired and it is no longer possible to pay it
```
```suggestion
    EXPIRED: Sorry, the notice has expired, it is not possible to pay it
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
+    ONGOING_SUBTITLE: If too much time has passed, please let us know!
+    REVOKED: Apologies, the Creditor Body has withdrawn this notice
+    EXPIRED: Apologies, the notice has expired and it is no longer possible to pay it
+    GENERIC_ERROR: Apologies, an unexpected error has occurred
```
```suggestion
    GENERIC_ERROR: Sorry, an unexpected error has occurred
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
+    REVOKED: Apologies, the Creditor Body has withdrawn this notice
+    EXPIRED: Apologies, the notice has expired and it is no longer possible to pay it
+    GENERIC_ERROR: Apologies, an unexpected error has occurred
+    GENERIC_ERROR_SUBTITLE: Try the payment again, or open a report for assistance.
```
```suggestion
    GENERIC_ERROR_SUBTITLE: You can try again, or contact us to get support.
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
+    EXPIRED: Apologies, the notice has expired and it is no longer possible to pay it
+    GENERIC_ERROR: Apologies, an unexpected error has occurred
+    GENERIC_ERROR_SUBTITLE: Try the payment again, or open a report for assistance.
+    PAYMENT_DUPLICATED: This notice has already been paid!
```
```suggestion
    PAYMENT_DUPLICATED: This notice has been already paid!
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
     PAYMENT_ONGOING_NOCANCEL_TIMEOUT: If you have already waited 30 minutes and the payment is still blocked, use the button below to send us a report.
     PAYMENT_ONGOING_CANCELLED: The previous attempt to pay this notice is being canceled, you can retry paying this notice after {{remainingMinutes}} minutes.
     PAYMENT_ONGOING_CANCELLED_TIMEOUT: The cancellation of this payment has beed requested more than 15 minutes ago but it is still blocked. Use the button below to send us a report.
+    contactECsubtitle: Contact the organisation for more information.
```
```suggestion
    contactECsubtitle: Contact the Institution for more information.
```

thisisjp commented on locales/en/index.yml:
```diff
@@ -1373,19 +1373,28 @@ wallet:
     PAYMENT_ONGOING_CANCELLED: The previous attempt to pay this notice is being canceled, you can retry paying this notice after {{remainingMinutes}} minutes.
     PAYMENT_ONGOING_CANCELLED_TIMEOUT: The cancellation of this payment has beed requested more than 15 minutes ago but it is still blocked. Use the button below to send us a report.
+    contactECsubtitle: Contact the organisation for more information.
+    assistanceLabel: "Error code for assistance:"
```
```suggestion
    assistanceLabel: "Use this code to get support:"
```

thisisjp commented on locales/it/index.yml:
```diff
@@ -1396,19 +1396,28 @@ wallet:
   errors:
-    PAYMENT_DUPLICATED: Questo avviso è già stato pagato!
+    TECHNICAL: Spiacenti, c’è un problema tecnico con questo avviso
+    DATA: Spiacenti, i codici inseriti non sono corretti
```
```suggestion
    DATA: Spiacenti, i dati dell’avviso non sono corretti
```

thisisjp commented on locales/it/index.yml:
```diff
@@ -1396,19 +1396,28 @@ wallet:
+    TECHNICAL: Spiacenti, c’è un problema tecnico con questo avviso
+    DATA: Spiacenti, i codici inseriti non sono corretti
+    EC: Spiacenti, l’Ente Creditore sta avendo problemi nella risposta
+    PAYMENT_ONGOING: C’è già un pagamento in corso, riprova tra qualche minuto
```
```suggestion
    PAYMENT_ONGOING: Il pagamento è già in corso, riprova tra qualche minuto
```

thisisjp commented on locales/it/index.yml:
```diff
@@ -1396,19 +1396,28 @@ wallet:
+    REVOKED: Spiacenti, l’Ente Creditore ha revocato questo avviso
+    EXPIRED: Spiacenti, l’avviso è scaduto e non è più possibile pagarlo
+    GENERIC_ERROR: Spiacenti, si è verificato un errore imprevisto
+    GENERIC_ERROR_SUBTITLE: Riprova il pagamento, oppure aprire una segnalazione per ricevere assistenza.
```
```suggestion
    GENERIC_ERROR_SUBTITLE: Prova di nuovo o contattaci per ricevere assistenza.
```

thisisjp commented on locales/it/index.yml:
```diff
@@ -1396,19 +1396,28 @@ wallet:
     PAYMENT_ONGOING_CANCELLED: È in corso l'annullamento del tuo tentativo di pagamento precedente, potrai riprovare a pagare l'avviso tra {{remainingMinutes}} minuti.
     PAYMENT_ONGOING_CANCELLED_TIMEOUT: L'annullamento di questo pagamento è stato richiesto più di 15 minuti fa ma risulta ancora bloccato. Utilizza il pulsante qui sotto per inviarci una segnalazione.
+    contactECsubtitle: Contatta l’Ente per maggiori informazioni.
+    assistanceLabel: "Codice di errore per l’assistenza:"
```
```suggestion
    assistanceLabel: "Per ricevere assistenza usa il codice:"
```

CrisTofani commented on ts/screens/wallet/payment/TransactionErrorScreen.tsx:
```diff
@@ -48,274 +56,261 @@ type Props = OwnProps &
-  };
-}
+  return (
+    <BaseScreenComponent>
```
No contextual help or back button is needed by design since the user can access to assistance by tapping the related CTA on the footer

CrisTofani commented on ts/screens/wallet/payment/TransactionErrorScreen.tsx:
```diff
@@ -48,274 +56,261 @@ type Props = OwnProps &
+    "blockedPaymentRptId",
+    RptIdFromString.encode(rptId)
+  );
+  BugReporting.showWithOptions(BugReporting.reportType.bug, [
```
This was on my TODO list, and then I forgot to add it, thanks for catching!

CrisTofani commented on ts/utils/payment.ts:
```diff
@@ -169,6 +172,121 @@ export const getErrorDescription = (
+  }
+};
+
+export const getErrorDescriptionV2 = (
+  error?: keyof typeof Detail_v2Enum
+): string | undefined => {
+  if (error === undefined) {
+    return undefined;
+  }
+
+  const errorMacro = getV2ErrorMacro(error);
+
+  switch (errorMacro) {
+    case "TECHNICAL":
+      return I18n.t("wallet.errors.TECHNICAL");
+    case "DATA":
+      return I18n.t("wallet.errors.DATA");
+    case "EC":
+      return I18n.t("wallet.errors.EC");
+    case "ONGOING":
+      return I18n.t("wallet.errors.PAYMENT_ONGOING");
+    case "REVOKED":
+      return I18n.t("wallet.errors.REVOKED");
+    case "EXPIRED":
+      return I18n.t("wallet.errors.EXPIRED");
+    case "DUPLICATED":
+      return I18n.t("wallet.errors.PAYMENT_DUPLICATED");
+    default:
+      return I18n.t("wallet.errors.GENERIC_ERROR");
+  }
+};
```
In the beginning I was thinking about the same refactoring then noticed that these localized labels was even translated in Deutsch language so I preferred not to rename them, anyway thinking it better now we can for sure refactorize it

CrisTofani commented on ts/screens/wallet/payment/TransactionErrorScreen.tsx:
```diff
@@ -48,274 +56,261 @@ type Props = OwnProps &
+  rptId: RptId,
+  onCancel: () => void
+): ScreenUIContents => {
+  const baseIconPath = "../../../../img/";
```
yes this is a const, but actually in this case is scoped only to be used by the current function, why to move on the entire file?

Undermaken replied:
because every time the function is called the const is created

CrisTofani replied:
f75e913 to solve

Undermaken commented on ts/screens/wallet/payment/TransactionErrorScreen.tsx:
```diff
@@ -48,277 +70,289 @@ type Props = OwnProps &
-  errorMessageSubtitle: {
-    textAlign: "center"
-  },
+const instabugTag = "payment-support";
```
Since it is used twice, could be a refactoring place it in the util file `utils/payment` and share between the components?

CrisTofani replied:
b3525ce to solve this

Undermaken commented on ts/screens/wallet/payment/TransactionErrorScreen.tsx:
```diff
@@ -2,19 +2,16 @@
  */
-import { differenceInMinutes } from "date-fns";
-import { Option, some } from "fp-ts/lib/Option";
+import { Option } from "fp-ts/lib/Option";
```
@CrisTofani 
Did you see this refactoring proposal? what do you think about it, could it be valid?

CrisTofani replied:
@Undermaken I missed it through all the messages 😅 

The refactoring looks good, just see one POF (talking about ui):

this code
```
  const subtitle = validError.fold(
    _ => (
      <H4 weight={"Regular"}>
        {I18n.t("wallet.errors.GENERIC_ERROR_SUBTITLE")}
      </H4>
    ),
    error => <ErrorCodeCopyComponent error={error} />
  );
```

would return the `ErrorCodeCopyComponent` on `UNCOVERED` cases which by design should have the GENERIC_ERROR_SUBTITLE text instead.

I'll give an attempt about these and propose another refactoring to you, in the next commit

Undermaken commented on ts/screens/wallet/payment/TransactionErrorScreen.tsx:
```diff
@@ -48,277 +71,272 @@ type Props = OwnProps &
+}): React.ReactElement => (
+  <View testID={"error-code-copy-component"}>
+    <H4 weight={"Regular"}>{I18n.t("wallet.errors.assistanceLabel")}</H4>
+    <H4 weight={"Bold"} testID={"error-code"}>
```
This PR LGTM
This is the only suggestion that I already tested (if you want to commit it directly)
```suggestion
    <H4 weight={"Bold"} testID={"error-code"}  style={{textAlign: "center"}}>
```

| now  | suggestion |
| ------------- | ------------- |
![now](https://user-images.githubusercontent.com/822471/134874820-b6dc3f60-30e8-4c19-889f-f42f48c745d9.png) | ![post](https://user-images.githubusercontent.com/822471/134874834-b45b60ad-a698-4d48-b5e8-c874be531c73.png)

Undermaken commented:
@CrisTofani Waiting green light from CI when it comes back

Undermaken commented:
@thisisjp for me is 🟢 
waiting your approval
